File: native-filters/src/main/java/com/facebook/imagepipeline/postprocessors/RoundAsCirclePostprocessor.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
package com.facebook.imagepipeline.postprocessors;

import android.graphics.Bitmap;
import com.facebook.cache.common.CacheKey;
import com.facebook.cache.common.SimpleCacheKey;
import com.facebook.imagepipeline.nativecode.NativeRoundingFilter;
import com.facebook.imagepipeline.request.BasePostprocessor;
import javax.annotation.Nullable;

/** Postprocessor that rounds a given image as a circle. */
public class RoundAsCirclePostprocessor extends BasePostprocessor {
  private static final boolean ENABLE_ANTI_ALIASING = true;

  private @Nullable CacheKey mCacheKey;
  private final boolean mEnableAntiAliasing;

  public RoundAsCirclePostprocessor() {
    this(ENABLE_ANTI_ALIASING);
  }

  public RoundAsCirclePostprocessor(boolean enableAntiAliasing) {
    mEnableAntiAliasing = enableAntiAliasing;
  }

  @Override
  public void process(Bitmap bitmap) {
    NativeRoundingFilter.toCircle(bitmap, mEnableAntiAliasing);
  }

  @Nullable
  @Override
  public CacheKey getPostprocessorCacheKey() {
    if (mCacheKey == null) {
      if (mEnableAntiAliasing) {
        mCacheKey = new SimpleCacheKey("RoundAsCirclePostprocessor#AntiAliased");
      } else {
        mCacheKey = new SimpleCacheKey("RoundAsCirclePostprocessor");
      }
    }
    return mCacheKey;
  }
}

File: native-filters/src/main/java/com/facebook/imagepipeline/postprocessors/RoundAsCirclePostprocessor.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.imagepipeline.postprocessors

import android.graphics.Bitmap
import com.facebook.cache.common.CacheKey
import com.facebook.cache.common.SimpleCacheKey
import com.facebook.imagepipeline.nativecode.NativeRoundingFilter
import com.facebook.imagepipeline.request.BasePostprocessor

/** Postprocessor that rounds a given image as a circle. */
class RoundAsCirclePostprocessor
@JvmOverloads
constructor(private val enableAntiAliasing: Boolean = ENABLE_ANTI_ALIASING) : BasePostprocessor() {

  private var cacheKey: CacheKey? = null

  override fun process(bitmap: Bitmap) {
    NativeRoundingFilter.toCircleFast(bitmap, enableAntiAliasing)
  }

  override fun getPostprocessorCacheKey(): CacheKey? {
    if (cacheKey == null) {
      cacheKey =
          if (enableAntiAliasing) {
            SimpleCacheKey("RoundAsCirclePostprocessor#AntiAliased")
          } else {
            SimpleCacheKey("RoundAsCirclePostprocessor")
          }
    }
    return cacheKey
  }

  companion object {
    private const val ENABLE_ANTI_ALIASING = true
  }
}
